Add ConcurrentBrokerRestartCountPerRack to RollingUpgradeConfig

ctrlaltluc opened the pull request:
## Description
Add `ConcurrentBrokerRestartCountPerRack` property to `RollingUpgradeConfig`, in order to allow multiple brokers to be restarted in parallel during a rolling upgrade.

This PR is a prerequisite for https://github.com/banzaicloud/koperator/pull/1001.

## Type of Change
- [x] New Feature

## Checklist
<!-- Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields. -->
- [x] I have read the [contributing guidelines](../CONTRIBUTING.md)
- [x] Existing issues have been referenced (where applicable)
- [x] I have verified this change is not present in other open pull requests
- [x] Functionality is documented
- [x] All code style checks pass
- [x] New code contribution is covered by automated tests
- [x] All new and existing tests pass

bartam1 commented on api/v1beta1/kafkacluster_types.go:
```diff
@@ -137,6 +137,16 @@ type RollingUpgradeConfig struct {
+	// This is a safe way to speed up the rolling upgrade. Note that for the rack distribution explained above, Cruise Control
+	// requires `com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareDistributionGoal` to be configured.
+	// +optional
+	ConcurrentBrokerRestartsAllowed int `json:"concurrentBrokerRestartsAllowed,omitempty"`
```
ConcurrentBrokerRestartsAllowed this name suggests that it type is bool for me.
I think it would be better to use something more descriptive name. Like ConcurrentBrokerRestartCountPerRack. What do you think?

ctrlaltluc replied:
Good idea, addressed in [6367359](https://github.com/banzaicloud/koperator/commit/6367359d714ed6f63d4e580edf2c78a20be506ce).

bartam1 reviewed (commented):
Just few comments. Otherwise LGTM.
Thank you for the contribution!

bartam1 commented on api/v1beta1/kafkacluster_types.go:
```diff
@@ -137,6 +137,16 @@ type RollingUpgradeConfig struct {
+	// This is a safe way to speed up the rolling upgrade. Note that for the rack distribution explained above, Cruise Control
+	// requires `com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareDistributionGoal` to be configured.
+	// +optional
+	ConcurrentBrokerRestartsAllowed int `json:"concurrentBrokerRestartsAllowed,omitempty"`
```
We can set a validation:
// +kubebuilder:validation:Minimum=1
// +kubebuilder:default=1

ctrlaltluc replied:
Added in [6367359](https://github.com/banzaicloud/koperator/commit/6367359d714ed6f63d4e580edf2c78a20be506ce).

ctrlaltluc commented:
@bartam1 Hi Marton, thanks a lot for the through reviews! Was on holiday until now. All your comments make sense, will have a look on your suggestions on the other PR too. Addressed all comments on this one.

ctrlaltluc commented:
Rebased again with master, ready for final review.

bartam1 reviewed (approved):
LGTM!

ctrlaltluc commented:
@panyuenlau this PR requires one more review to be merged. Can you help please? Asking you as you already had a look on https://github.com/banzaicloud/koperator/pull/1001 and this one is a prerequisite for that. Thanks! :)

panyuenlau reviewed (approved):
reviewed the main implementation - looks great

ctrlaltluc commented:
Rebased again with master (no other changes) in hopes to have it merged and [there is an integration test failing](https://github.com/banzaicloud/koperator/actions/runs/5681154322/job/15396691867?pr=1002#step:11:91). First time I see it failing here.
Probably a flaky one? Is there any way to re-run it?

```
[FAILED] [120.019 seconds]
KafkaClusterNodeportExternalAccess [JustBeforeEach] when NodePortExternalIP is configured reconciles the service successfully
  [JustBeforeEach] /home/runner/work/koperator/koperator/controllers/tests/kafkacluster_controller_externalnodeport_test.go:78
  [It] /home/runner/work/koperator/koperator/controllers/tests/kafkacluster_controller_externalnodeport_test.go:147
```

panyuenlau commented:
> Rebased again with master (no other changes) in hopes to have it merged and [there is an integration test failing](https://github.com/banzaicloud/koperator/actions/runs/5681154322/job/15396691867?pr=1002#step:11:91). First time I see it failing here. Probably a flaky one? Is there any way to re-run it?
> 
> ```
> [FAILED] [120.019 seconds]
> KafkaClusterNodeportExternalAccess [JustBeforeEach] when NodePortExternalIP is configured reconciles the service successfully
>   [JustBeforeEach] /home/runner/work/koperator/koperator/controllers/tests/kafkacluster_controller_externalnodeport_test.go:78
>   [It] /home/runner/work/koperator/koperator/controllers/tests/kafkacluster_controller_externalnodeport_test.go:147
> ```

Context: The controller test failure "spec.ports[0].nodePort: Invalid value: 31125: provided port is already allocated" is seen from time to time in GHA

My guess: this is *probably* because the underlying node that GHA uses to run the tests happens to be the same one that was used in previous run, and GHA's infrastructure didn't get to clean up the allocated port in time

ctrlaltluc commented:
> > Rebased again with master (no other changes) in hopes to have it merged and [there is an integration test failing](https://github.com/banzaicloud/koperator/actions/runs/5681154322/job/15396691867?pr=1002#step:11:91). First time I see it failing here. Probably a flaky one? Is there any way to re-run it?
> > ```
> > [FAILED] [120.019 seconds]
> > KafkaClusterNodeportExternalAccess [JustBeforeEach] when NodePortExternalIP is configured reconciles the service successfully
> >   [JustBeforeEach] /home/runner/work/koperator/koperator/controllers/tests/kafkacluster_controller_externalnodeport_test.go:78
> >   [It] /home/runner/work/koperator/koperator/controllers/tests/kafkacluster_controller_externalnodeport_test.go:147
> > ```
> 
> Context: The controller test failure "spec.ports[0].nodePort: Invalid value: 31125: provided port is already allocated" is seen from time to time in GHA
> 
> My guess: this is _probably_ because the underlying node that GHA uses to run the tests happens to be the same one that was used in previous run, and GHA's infrastructure didn't get to clean up the allocated port in time

@panyuenlau understood. 

We thought it only happens in the Adobe fork.
We attempted to fix this in https://github.com/adobe/koperator/pull/67, by explicitly deleting the nodeports. I thought it was related to `envtest` [not doing garbage collection](https://cluster-api.sigs.k8s.io/developer/testing.html) (when `KafkaCluster` is deleted, even if it is a parent of those nodeports, the nodeports are not deleted):
```
One common example is garbage collection; because there are no controllers monitoring built-in resources, objects do not get deleted, even if an OwnerReference is set up; as a consequence, usually test implements code for cleaning up created objects.
```

But it seems [explicitly deleting the nodeports](https://github.com/adobe/koperator/pull/67) does not fix it either. We still saw it afterwards, on occasion.
